### ISSUE TITLE
Planning our next major version increment

### ISSUE THREAD

**Issue by Zac-HD** [HypothesisWorks/hypothesis#2709]
TLDR; next major release will be Hypothesis 6.0.0 - WIP@https://github.com/HypothesisWorks/hypothesis/compare/master...Zac-HD:hypothesis-6 - as soon as:

- [x] @Zalathar / #2733 our function-scoped fixture warning can be disabled with the `suppress_health_check` setting
- @Zac-HD / #2705 we have a codemod for `ip_addr_strings()` -> `ip_addresses().map(str)`
  *I've decided to leave this until later; correctly replacing imports turned out to be tricky and it's way lower priority than the positional-to-keyword argument codemod.*

After which development, including of further codemods, will continue as usual 😁 

-------------

For the last few years, we've released a new major-version in early January, using it to clear out our deprecations and notably to drop Python 2 (🎉).  This issue is to invite discussion, from @HypothesisWorks/hypothesis-python-contributors and anyone interested in the wider community, about both our next major version and our policy around them.

**The next major version** will remove or alter seven deprecated APIs, and change the calling convention of a further 46 functions to use keyword-only arguments (via `@deprecated_posargs`).  In all cases these were either silently broken, or are older than our six-month minimum deprecation period and can therefore be removed.  However, I would like to ship codemods to automate the update (#2705) for at least the kwonly args change before converting them to a hard error; likely meaning later in January.

**I like our major version policy**, even if [semver under-specifies "breaking change"](https://snarky.ca/why-i-dont-like-semver/) (I think we take this more seriously than many other libraries I use).  ~~An annual release cadence does suggest that [calver might be a better choice](https://calver.org/), so I propose that we use `pip`-style "YY.minor.micro" versioning.  Specifically, we would release version `21.0.0`, remain semver-compliant, and clearly communicate [age in libyears](https://libyear.com/) rather than giving an impression of API instability.~~  (thanks @sobolevn!)

Thoughts?

**Comment by sobolevn** [comment]
Hi!

>  An annual release cadence does suggest that calver might be a better choice, so I propose that we use pip-style "YY.minor.micro" versioning.

I have some experience with calver as both user and maintainer and I want to share it.

The thing with `An annual release cadence does suggest that calver might be a better choice` is that this just happens to be the case right now and can be changed in the future. 

Moreover, it has some clear disadvantages:

- Switching to calver will not allow us to ship a breaking change in the middle of the year even if we need to. 
- The release cycle might also be larger than just one year. For example, there might be a year without a major release. This will leave users with an impression of using an outdated library. Like using `22.1.0` in 2024. `pipenv` is the most canonical example here.

What is even more important is that switching to calver is not something that can be easily undone.

In my opinion, calver is a great thing for libraries like `pytz` or security-related ones or any other libs that **need** to indicate when they were released. Which is not the case for `hypothesis` IMHO.

Anyway, thanks for the opportunity to discuss it publicly! 👍 

P.S. Happy holidays! 🎅

**Comment by Zac-HD** [comment]
Thanks, that's very valuable feedback!  My updated proposal then is that we continue our recent pattern, which I'd summarise as "major version updates as the core team judges a good idea; and Zac usually proposes on early in the new year".

I'd be perfectly happy skipping a year if there weren't many deprecations; and in fact I *would* skip this year if it wasn't for the keyword-only calling convention change... using real kwonly args instead of our compatibility shim will be a somewhat nicer user experience, and *far* easier for tools to deal with (which I care about because good interop = good UX).

**Comment by Zac-HD** [comment]
Summaries of previous major releases for reference:

**1.0**: graduated from https://0ver.org (🎉)  
**2.0**: cleaned up lots of legacy stuff from 1.x (~9 months), some without prior notice  
**3.0**: all-new `conjecture` backend, no deprecations as it only took a few weeks  
**4.0**: dropped three years' of deprecations and old-style features  
**5.0**: finalised many small API style deprecations, removed most py2 compat stuff  
**6.0** *(proposed)*: require kwonly args, drop a few remaining deprecations

While we've added a *lot* of nice new things since 4.0, thanks to our API style guide there's very little that we want to deprecate.  Specifically, presently deprecated and scheduled for removal in Hypothesis 6.0:

- Passing the `from_model(model=...)` as a keyword rather than positional argument (like `st.builds()`)
- Using Pytest function-scoped fixtures with `@given`
- Passing `st.complex_numbers(min_magnitude=None)` instead of `min_magnitude=0`, has codemod
- Provisional `ipN_addr_strings()` strategies rather than `st.ip_addresses(v=N).map(str)`, codemod planned
- `st.randoms(use_true_random=False)` will be the default behaviour, `=None` becomes an error
- Calling `st.register_type_strategy()` on a parametrised generic type becomes an error (never worked)
- And finalising the change to keyword-only calling conventions for many functions

And this is around [one-third as many as 5.0](https://hypothesis.readthedocs.io/en/latest/changes.html#v5-0-0) and [one-fifth as many as 4.0](https://hypothesis.readthedocs.io/en/latest/changes.html#v4-0-0).

**Comment by Zalathar** [comment]
I'm thinking about adding a health check for function-scoped fixtures, so that rare intentional uses (e.g. #2731) can continue to opt-out when the warning becomes an error.

I haven't worked out the implementation details yet (it's annoyingly non-trivial), but I'll probably want to coordinate landing that as part of the major version bump, for a relatively smooth transition.
